File: 백준/Gold/13703. 물벼룩의 생존확률/물벼룩의 생존확률.py
# 물벼룩의 생존확률
import sys
input = sys.stdin.readline

def solve(loc, n):

    if dp[loc][n] != -1:
        return dp[loc][n]
    else:
        dp[loc][n] = 0

    if loc == 0: # 수면에 닿았다면 2^(남은 횟수) 리턴
        dp[loc][n] = 2**n
        return dp[loc][n]

    if n == 0:
        dp[loc][n] = 0
        return 0
    
    dp[loc][n] = solve(loc+1, n-1) + solve(loc-1, n-1)

    return dp[loc][n]

K, N = map(int,input().split())
dp = [[-1] * 64 for _ in range(128)]

print(2**N - solve(K, N))

File: 백준/Gold/13703. 물벼룩의 생존확률/물벼룩의 생존확률.java
import java.io.*;
import java.util.*;

public class Main {

    static long[][] dp = new long[128][64];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int K = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());

        for (int i = 0; i < 128; i++) {
            Arrays.fill(dp[i], -1);
        }

        System.out.println(((long)  1 << N) - solve(K, N));
    }

    static long solve(int loc, int n) {

        if (dp[loc][n] != -1) {
            return dp[loc][n];
        } else {
            dp[loc][n] = 0;
        }

        if (loc == 0) {
            dp[loc][n] = (long) 1 << n;
            return dp[loc][n];
        }

        if (n == 0) {
            dp[loc][n] = 0;
            return dp[loc][n];
        }

        dp[loc][n] = solve(loc -1, n-1) + solve(loc+1, n-1);

        return dp[loc][n];
    }
}
